### PR TITLE
Move source removal to Module instead of Lexer

### PR DESCRIPTION
**Pull request by nicolas-grekas**
See comments here:
https://github.com/twigphp/Twig/pull/2164/files#r82133239

ping @stof @LionsAd

(side note: should getDebugInfo be conditionally populated also?)

### CODE REVIEW

**Comment by stof** [comment]
:+1: 

This means that ``getSource`` should not be changed to abstract in master anymore though

**Comment by LionsAd** [comment]
👍 

Looks great to me, I think it would make sense to make getDebugInfo() conditional on isDebug(), too. (as the name says)

And yes both methods could no longer be abstract in 2.x/master.

File: lib/Twig/Node/Module.php [hunk 1]
```diff
@@ -94,7 +94,9 @@ protected function compileTemplate(Twig_Compiler $compiler)
 
         $this->compileDebugInfo($compiler);
 
-        $this->compileGetSource($compiler);
+        if ($compiler->getEnvironment()->isDebug()) {
+            $this->compileGetSource($compiler);
+        }
```
Comment (fabpot) [inline lib/Twig/Node/Module.php]: I would always compile the `getSource()` method but make it return an empty string if debug is disabled.

File: lib/Twig/Lexer.php [hunk 2]
```diff
@@ -136,7 +136,7 @@ public function tokenize($code, $filename = null)
             mb_internal_encoding($mbEncoding);
         }
 
-        return new Twig_TokenStream($this->tokens, $this->filename, $this->env->isDebug() ? $code : '');
+        return new Twig_TokenStream($this->tokens, $this->filename, $code);
```
Comment (fabpot) [inline lib/Twig/Lexer.php]: I would revert this totally and remove the `$code` from `Twig_TokenStream` (this is purely internal and nobody should rely on this).

**Comment by fabpot** [comment]
`getDebugInfo()` must always be populated for two reasons: first, it does not add that much code, then, it is used to give better exception messages (which is also useful in production).

**Comment by nicolas-grekas** [comment]
Closing in favor of #2173

**Pull request closed without merging**